From a8e80cd46f9f5b24d08c4b33a9e2517777dbf85b Mon Sep 17 00:00:00 2001
From: ylevavi <yehuda.levavi@tandemg.com>
Date: Mon, 11 Oct 2021 12:41:42 +0300
Subject: [PATCH] [SW-42060] preboot: gaudi2: add tpm flow

Add TPM flow to Gaudi2 preboot:
	- Send STARTUP command
	- Send NV_READ command to get the Device ID from the TPM
	- Read Device ID from EFUSE.
	- Compare the IDs:
		- equal: continue.
		- different: print to serial log and report to LKD.
		- efuse empty: print warning to serial log and report to
		  LKD (efuse write will be handled by fw-loader agent
		  with gaudi2-write-tpm-devid image).

Change-Id: Ic6c00dde14fa65e4739ba9bc57a21b2d336ceaf3
Signed-off-by: ylevavi <yehuda.levavi@tandemg.com>
---

diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index a3e1f99..e1bf13b 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -389,6 +389,7 @@
 __weak void preboot_read_test_mode(void) {}
 __weak void preboot_reset_init(void) {}
 __weak void preboot_cond_arc0_reset(void) {}
+__weak void preboot_tpm_auth(void) {}
 
 void preboot_main(void)
 {
@@ -461,6 +462,8 @@
 
 	preboot_clear_wd_gpio();
 
+	preboot_tpm_auth();
+
 	module_handle_cmds(cmd_ops, ARRAY_SIZE(cmd_ops));
 }
 
diff --git a/arch/gaudi2/common/asic.c b/arch/gaudi2/common/asic.c
index 4f3061c..cab2480 100644
--- a/arch/gaudi2/common/asic.c
+++ b/arch/gaudi2/common/asic.c
@@ -19,6 +19,18 @@
 #include <pci.h>
 #include <sha.h>
 #include <efuse.h>
+#include <tpm2.h>
+#include <lzo/unaligned.h>
+#include <boot_if.h>
+
+#define TPM_NV_DEV_SN_ADDRESS	0x01c10100
+
+struct tpm_dev_id {
+	union {
+		u32 efuse_data[2];
+		u64 tpm_data;
+	};
+};
 
 #define HBM_BAR_PF_NUM (1 << 4)
 
@@ -248,3 +260,63 @@
 {
 	hl_fuse_print_hbm();
 }
+
+static int preboot_get_efuse_dev_id(struct tpm_dev_id *efuse_dev_id_ret)
+{
+	int ret;
+
+	ret = preboot_fuse_read(EFUSE_DEV_ID_BACKUP_OFFSET_1, &efuse_dev_id_ret->efuse_data[0]);
+	if (ret)
+		return ret;
+
+	ret = preboot_fuse_read(EFUSE_DEV_ID_BACKUP_OFFSET_2, &efuse_dev_id_ret->efuse_data[1]);
+	if (ret)
+		return ret;
+
+	return 0;
+}
+
+void preboot_tpm_auth(void)
+{
+	int ret;
+	u64 tpm_dev_id = 0;
+	struct tpm_dev_id efuse_dev_id;
+
+	efuse_dev_id.tpm_data = 0x0;
+
+	ret = tpm2_init();
+	if (ret)
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+
+	ret = tpm2_startup(TPM2_SU_CLEAR);
+	if (ret) {
+		preboot_puts_hex("TPM: startup failed. ret = ", ret, true);
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	}
+
+	ret = tpm2_nv_read(TPM_NV_DEV_SN_ADDRESS, &tpm_dev_id, sizeof(tpm_dev_id));
+	if (ret) {
+		preboot_puts_hex("TPM: NV_READ failed. ret = ", ret, true);
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	}
+
+	/* tpm device id returns as Big Endian, need to convert */
+	tpm_dev_id = get_unaligned_be64(&tpm_dev_id);
+	preboot_puts_hex64("TPM device ID: ", tpm_dev_id, true);
+
+	ret = preboot_get_efuse_dev_id(&efuse_dev_id);
+	if (ret) {
+		preboot_puts_hex("TPM flow: efuse read failed. ret = ", ret, true);
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	} else {
+		preboot_puts_hex64("EFUSE device ID: ", efuse_dev_id.tpm_data, true);
+	}
+
+	if (efuse_dev_id.tpm_data == 0) {
+		preboot_puts("TPM flow: error: efuse device ID not programed\n");
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	} else if (efuse_dev_id.tpm_data != tpm_dev_id) {
+		preboot_puts("TPM flow: error: efuse device ID is different than TPM device ID\n");
+		cpu_boot_err_set(CPU_BOOT_ERR_TPM_FAIL);
+	}
+}
diff --git a/arch/gaudi2/common/tpm2.h b/arch/gaudi2/common/tpm2.h
index 33a8730..c0119bd 100644
--- a/arch/gaudi2/common/tpm2.h
+++ b/arch/gaudi2/common/tpm2.h
@@ -63,5 +63,6 @@
 unsigned int get_sclk_out_pad_sel_tpm(void);
 void spi_ssi_cs1_get(unsigned int *mask, unsigned int *data_offset,
 		     unsigned int *out_en_offset);
+void preboot_tpm_auth(void);
 
 #endif /* _HL_TPM2_H_ */
diff --git a/lib/common/efuse.h b/lib/common/efuse.h
index b3b1f61..2a5cd15 100644
--- a/lib/common/efuse.h
+++ b/lib/common/efuse.h
@@ -12,6 +12,8 @@
 
 #define EFUSE_READ_RANGE_ALL		0x3F80
 #define EFUSE_PRODUCT_ID_OFFSET		0x1
+#define EFUSE_DEV_ID_BACKUP_OFFSET_1	0x38
+#define EFUSE_DEV_ID_BACKUP_OFFSET_2	0x39
 #define EFUSE_SYSCFG_OFFSET             0x3A
 
 int preboot_fuse_init(void);
